### PR TITLE
fix global refinement

### LINKED ISSUE #4524: use cell flag communication for initial refinement
See discussion in #4513

This is what we should do, I think:
https://github.com/geodynamics/aspect/blob/048eeedbbdaed04b7a342f0faa47556c094b6e97/source/simulator/core.cc#L1616-L1644

@jdannberg do you think you can try that? If not, I can prepare it as well.

_Originally posted by @tjhei in https://github.com/geodynamics/aspect/issues/4513#issuecomment-1062027271_

### PR DESCRIPTION
**Pull request by tjhei**
Requesting global refinement when running with many MPI ranks often
resulted in cells that we forget to refine.

fixes #4524
also see #4513
also see various forum posts about continental_extension cookbook with
16 MPI ranks and others.

continental_extension:
```
$ grep Number *ranksafter-16-ranks:Number of active cells: 3,200 (on 4 levels)
before-16-ranks:Number of active cells: 3,056 (on 4 levels)
before-16-ranks:Number of degrees of freedom: 103,689 (25,122+3,201+12,561+12,561+12,561+12,561+12,561+12,561)
before-16-ranks:Number of mesh deformation degrees of freedom: 6402
before-8-ranks:Number of active cells: 3,200 (on 4 levels)
before-8-ranks:Number of degrees of freedom: 107,649 (26,082+3,321+13,041+13,041+13,041+13,041+13,041+13,041)
before-8-ranks:Number of mesh deformation degrees of freedom: 6642

after-16-ranks:Number of active cells: 3,200 (on 4 levels)
after-16-ranks:Number of degrees of freedom: 107,649 (26,082+3,321+13,041+13,041+13,041+13,041+13,041+13,041)
after-16-ranks:Number of mesh deformation degrees of freedom: 6642
after-8-ranks:Number of active cells: 3,200 (on 4 levels)
after-8-ranks:Number of degrees of freedom: 107,649 (26,082+3,321+13,041+13,041+13,041+13,041+13,041+13,041)
after-8-ranks:Number of mesh deformation degrees of freedom: 6642

```

### CODE REVIEW

**Comment by tjhei** [comment]
Looks like no test captures this issue. I am hesitant to add a test with many MPI ranks, though.

File: source/simulator/core.cc [hunk 1]
```diff
@@ -1954,6 +1930,9 @@ namespace aspect
                 cell->set_refine_flag ();
 
             mesh_refinement_manager.tag_additional_cells ();
+
+            exchange_refinement_flags();
```
Comment (bangerth) [inline source/simulator/core.cc:1934]: So this is the place where we forgot it!

**Review by bangerth — COMMENTED** [review]
Looks good to me. Nice detective work.

I think it's ok to have a test with 16 processes running if you make it real short -- e.g., one time step, no substantial solve. Would you mind adding one?

**Comment by tjhei** [comment]
Adding a test with 16 ranks is problematic on GitHub actions with 2 real cores, right @gassmoeller ?

**Comment by gassmoeller** [comment]
Running with many MPI cores in tests caused problems in the past, but I think they are all solved by now. For reference we now set the following Open MPI environment variables inside the docker container that (hopefully) solve all issues we had:
```
ENV OMPI_MCA_btl_base_warn_component_unused=0
ENV OMPI_MCA_mpi_yield_when_idle=1
ENV OMPI_MCA_rmaps_base_oversubscribe=1
ENV OMPI_ALLOW_RUN_AS_ROOT=1
ENV OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1
```

So I would say feel free to add a test with many MPI ranks (we already have test with up to 4, so more than the 2 that are available for github actions).

**Comment by tjhei** [comment]
I added a test with 8 ranks (this is as small as I was able to get it)

**Review by bangerth — APPROVED** [review]
Thank you!

**Pull request merged**